torchtitan/datasets/hf_datasets.py

```diff
@@ -107,36 +108,71 @@ def __init__(
 
         self.dataset_name = dataset_name
         self._data = split_dataset_by_node(ds, rank, world_size)
+        self._sample_idx = 0
         self._tokenizer = tokenizer
+        self._all_tokens: List[int] = []
         self.seq_len = seq_len
         self.infinite = infinite
+        self._first_iter = True
 
     def __iter__(self):
         max_buffer_token_len = 1 + self.seq_len
-        all_tokens: List[int] = []
 
         while True:
-            for sample in iter(self._data):
+            # Skip samples for the first iteration
+            data_iter = (
+                self._get_skipped_data_iter(self._sample_idx)
+                if self._first_iter
+                else iter(self._data)
+            )
+            self._first_iter = False
+
+            for sample in data_iter:
                 sample_text = sample["text"]
                 sample_tokens = self._tokenizer.encode(sample_text, bos=True, eos=True)
-                all_tokens.extend(sample_tokens)
+                self._all_tokens.extend(sample_tokens)
+                self._sample_idx += 1
 
-                while len(all_tokens) >= max_buffer_token_len:
-                    x = torch.LongTensor(all_tokens[:max_buffer_token_len])
+                while len(self._all_tokens) >= max_buffer_token_len:
+                    x = torch.LongTensor(self._all_tokens[:max_buffer_token_len])
                     # update tokens to the remaining tokens
-                    all_tokens = all_tokens[max_buffer_token_len:]
+                    self._all_tokens = self._all_tokens[max_buffer_token_len:]
                     input = x[:-1]
                     label = x[1:]
                     yield input, label
+
             if not self.infinite:
                 logger.warning(f"Dataset {self.dataset_name} has run out of data.")
                 break
             else:
+                # Reset offset for the next iteration
+                self._sample_idx = 0
                 logger.warning(
                     f"Dataset {self.dataset_name} is being re-looped. "
                     "Loss related metrics might be misleading."
                 )
 
+    def _get_skipped_data_iter(self, skip_samples):
+        if isinstance(self._data, IterableDataset):
+            it = iter(self._data)
+            # Naively iterate through the samples as dataset length is unknown
+            for _ in range(skip_samples):
+                next(it)
+            return it
+
+        # Map-style dataset
+        # As skipping to the end throws an error, return an empty iterator
+        if skip_samples == len(self._data):
```

Comment:
Does this happen at end-of-epoch? In this case should we just restart the iterator instead of returning empty iterator?

Reply:
@andrewkho Yes, this is for end-of-epoch. I structured this method to take care of creating a skipped iterator. The caller takes care of restarting if self.infinite is set to true.